StoreUtil use page cache to move and rename files

burqen opened the pull request:
**What?**
Use page cache as well as file system to move and delete files when copy store in HA.

**Why?**
When using a different storage device (blockdevice) for record files, only the page cache knows how to move and delete those files.

Changes:
`StoreUtil` use page cache to move and delete files.
`BranchedDataMigrator` use page cache to move files.
Added `pathToFileAfterMove` method to `FileUtils`.

Only last commit in PR is new, rebased on top of #8031

Commit:
StoreUtil use page cache to move and rename files

What?
Use page cache as well as file system to move and delete files when copy store in HA.

Why?
When using a different storage device (blockdevice) for record files, only the page cache knows how to move and delete those files.

Changes:
StoreUtil use page cache to move and delete files.
BranchedDataMigrator use page cache to move files.
Added pathToFileAfterMove method to FileUtils.

chrisvest commented:
I suggest you also `rename {community/kernel/src/main/java/org/neo4j/kernel/impl/util => enterprise/ha/src/main/java/org/neo4j/kernel/ha/store}/StoreUtil.java` because it's only used by the branched data handling machinery.

chrisvest commented:
Once #8031 is merged, this one will need a rebase. /cc @lutovich

Commit:
Move StoreUtil to enterprise ha store package

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/BranchedDataMigrator.java:
```diff
@@ -67,6 +71,12 @@ private void migrateBranchedDataDirectoriesToRootDirectory()
             try
             {
                 FileUtils.moveFile( oldBranchedDir, targetDir );
+                Iterable<FileHandle> handles = pageCache.streamFilesRecursive( oldBranchedDir )::iterator;
```
You have already exactly the same thing in StoreUtil

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -63,9 +66,30 @@ public static File newBranchedDataDir( File storeDir )
+    {
+        FileUtils.deleteRecursively( storeDir );
+        Iterable<FileHandle> handles = pageCache.streamFilesRecursive( storeDir )::iterator;
+        for ( FileHandle handle : handles )
```
i would extract handle.delete into consumer that wraps it into UncheckedIOException and use plain stream pipeline

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -63,9 +66,30 @@ public static File newBranchedDataDir( File storeDir )
+        Iterable<FileHandle> handles = pageCache.streamFilesRecursive( storeDir )
+                .filter( fh -> DEEP_STORE_FILE_FILTER.accept( fh.getFile() ) )::iterator;
+        for ( FileHandle handle : handles )
+        {
```
here the same thing i would extract for loop content into consumer and pass it into same stream pipeline isntead

burqen replied:
Problem is `rename` throws IOException, how to handle that in stream?

burqen replied:
Well, `Unchecked...`.

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -84,6 +108,35 @@ public static File getBranchedDataDirectory( File storeDir, long timestamp )
+            return new File[0];
+        }
+
+        return storeDir.listFiles( file -> {
```
we can extract this filter into a private static field to be more close to his close friend `DEEP_STORE_FILE_FILTER`

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -84,6 +108,35 @@ public static File getBranchedDataDirectory( File storeDir, long timestamp )
+
+    private static boolean isBranchedDataRootDirectory( File file )
+    {
+        return file.isDirectory() && file.getName().equals( BRANCH_SUBDIRECTORY );
```
flip this equals please

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -5,55 +5,58 @@
+        for ( FileHandle handle : handles )
         {
-            return new File[0];
+            handle.delete();
```
same as below (extraction to consumer instead)

MishaDemianenko reviewed (commented):
Have couple of comments, most of them are minor things

chrisvest commented:
There are also relevant test failures, plus it obviously needs a rebase once #8031 is merged.

chrisvest commented:
@burqen #8031 has now been merged.

Commit:
Address PR comments

- Extract consumers for FileHandle delete and rename to make it possible
to use in stream pipeline.
- When, in StoreUtil, moving files between directories with PageCache if
directory to move from does not exist because it has already been moved with
file system we don't break down.
- Use Apache commons instead of homebrewed method to check if string is numeric.

burqen commented:
@chrisvest @MishaDemianenko Comments addressed and PR rebased.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/FileHandle.java:
```diff
@@ -30,6 +34,46 @@
+     * <p>
+     * Possible IOException caused by fileHandle.delete() is wrapped in UncheckedIOException
+     */
+    Consumer<FileHandle> handleDelete = fh -> {
```
`HANDLE_DELETE` would be a better name

lutovich commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/store/StoreUtil.java:
```diff
@@ -0,0 +1,159 @@
+        return file.getPath().contains( BRANCH_SUBDIRECTORY );
+    }
+
+    private static boolean isNumerical( String string )
```
`StringUtils.isNumeric()` can be used instead

lutovich reviewed (approved):
LGTM, two really minor comments

burqen commented:
@lutovich Comments addressed.

burqen commented:
Need to confirm that HA-robustness is not broken by this PR.

chrisvest commented:
@burqen Restarted your LInux build because it failed when it somehow couldn't download artefacts from the UK maven mirror.

MishaDemianenko commented:
@burqen approved. LGTM on green!

Commit:
SwitchToSlaveTest fix test failure

burqen commented:
`TransactionConstraintsIT` test failure seem unrelated.
Pushing a fix for `SwithToSlaveTest`
